### ISSUE TITLE
MOPD top-k same-vocab KL materializes a full-vocab student log-softmax that cancels out of the loss

### ISSUE THREAD

**Issue by tianyi-zhang-02** [NVIDIA-NeMo/RL#3272]
## Summary

`_topk_same_vocab_kl` in `nemo_rl/algorithms/loss/loss_functions.py` computes the student side via `vocab_parallel_full_log_softmax` — materializing a full-vocab `[B, T, V]` fp32 log-prob tensor (autograd-tracked, and all-gathered across TP ranks) — then uses only the K teacher top-k columns and renormalizes within that subset:

```python
student_log_probs = vocab_parallel_full_log_softmax(student_logits, T, tp_group=tp_group)
student_gathered = student_log_probs[..., topk_idx]
student_log_probs_k = student_gathered - torch.logsumexp(student_gathered, dim=-1, keepdim=True)
```

The method's own docstring notes *"the full-vocab partition function cancels"* — and that cancellation means the full-vocab tensor is mathematically unnecessary:

```
log_softmax_V(x/T)[idx] − lse_K(log_softmax_V(x/T)[idx])
  = (x/T − lse_V)[idx] − (lse_K(x/T[idx]) − lse_V)
  = log_softmax_K(x/T[idx])
```

Since `select_teacher_topk_indices` returns a shared 1-D `[K]` column set, the student side reduces to slicing K columns of the raw logits and a `log_softmax` over K.

## Numerical evidence (fp32, both KL directions)

- Forward loss: **bit-identical** (abs diff 0.0e+00 forward KL; 6e-08 reverse KL)
- Gradients w.r.t. `student_logits`: max abs diff **1.9e-09** against grad magnitudes ~1.8e-02 (~1e-7 relative)

## Memory / comms impact

Per sample at `T=4096`, `V=151936` (Qwen tokenizer), `K=64`, fp32:

| | dominant intermediate |
|---|---|
| current | `[T, V]` = **2.32 GiB** (plus the autograd-saved copy) |
| proposed | `[T, K]` = **1.0 MiB** (~2400x smaller) |

At TP > 1 the current path additionally all-gathers the full vocab axis to every rank (`gather_from_tensor_model_parallel_region`); the proposed path only needs the K columns — the cross-tokenizer student path already does exactly this via `gather_logits_at_global_indices`, so the TP-aware machinery exists in-tree.

## Scope

- Only the **top-k** path. `_direct_full_vocab_kl` genuinely needs the full vocab (its KL spans V) and is not part of this proposal.
- I'm aware MOPD is under active development (#2780 / roadmap), so I'd like to coordinate rather than surprise-PR into moving code. Happy to send the change with an equivalence regression test (forward + grad, both KL directions) if the shape looks right to you.

cc @terrykong @yuki-97

**Comment by tianyi-zhang-02** [comment]
cc @avenkateshha @yfw as the authors of the xtoken/MOPD loss paths (#2508, #2780) — you're best placed to sanity-check the cancellation argument, since it leans on the method's own docstring.

Also correcting my framing above: #2780 has merged, so this targets settled mainline code rather than in-flight work. A prototype branch with the change plus equivalence tests (forward and gradients vs. the previous formulation, at multiple temperatures) is ready to go as a PR if the shape looks right to you.

**Comment by tianyi-zhang-02** [comment]
Opened #3314 with the change to make review concrete — the `vocab_parallel_gather_columns` helper (reusing `gather_logits_at_global_indices` for the TP path), the `_direct_topk_kl` rewrite, and equivalence tests asserting forward values and gradients match the previous formulation at multiple temperatures. Happy to adjust or close if you'd prefer a different shape.
